weston-ivi-shell/src/ivi-controller.c

```diff
@@ -1095,80 +1131,163 @@ controller_screen_remove_layer(struct wl_client *client,
-        wl_container_of(listener, l, listener);
-    char *filename = l->filename;
+        wl_container_of(listener, l, frame_listener);
 
     struct weston_output *output = data;
     int32_t width = 0;
     int32_t height = 0;
     int32_t stride = 0;
-    uint8_t *readpixs = NULL;
+    uint32_t *readpixs = NULL;
+    uint32_t shm_format;
+    int fd;
+    size_t size;
+    pixman_format_code_t format = output->compositor->read_format;
 
     --output->disable_planes;
-    wl_list_remove(&listener->link);
 
     width = output->current_mode->width;
     height = output->current_mode->height;
-    stride = width * (PIXMAN_FORMAT_BPP(output->compositor->read_format) / 8);
+    stride = width * (PIXMAN_FORMAT_BPP(format) / 8);
+    size = stride * height;
 
-    readpixs = malloc(stride * height);
-    if (readpixs == NULL) {
-        weston_log("fails to allocate memory\n");
-        free(l->filename);
-        free(l);
-        return;
+    fd = create_screenshot_file(size);
+    if (fd < 0) {
+        weston_log("screenshot: failed to create file of %zu bytes: %m\n",
+                   size);
+        ivi_screenshot_send_error(l->screenshot, IVI_SCREENSHOT_ERROR_IO_ERROR,
+                                  "failed to create screenshot file");
+        goto err_fd;
+    }
+
+    readpixs = mmap(NULL, size, PROT_READ | PROT_WRITE, MAP_SHARED, fd, 0);
+    if (readpixs == MAP_FAILED) {
+        weston_log("screenshot: failed to mmap %zu bytes: %m\n", size);
+        ivi_screenshot_send_error(l->screenshot, IVI_SCREENSHOT_ERROR_IO_ERROR,
+                                  "failed to create screenshot");
+        goto err_mmap;
     }
 
-    output->compositor->renderer->read_pixels(
-            output,
-            output->compositor->read_format,
-            readpixs,
-            0,
-            0,
-            width,
-            height);
+    if (output->compositor->renderer->read_pixels(output, format, readpixs,
+                                                  0, 0, width, height) < 0) {
+        ivi_screenshot_send_error(
+            l->screenshot, IVI_SCREENSHOT_ERROR_NOT_SUPPORTED,
+            "screenshot of given output is not supported by renderer");
+        goto err_readpix;
+    }
+
+    if (output->compositor->capabilities & WESTON_CAP_CAPTURE_YFLIP)
+        flip_y(stride, height, readpixs);
+
+    // map to shm buffer format
+    switch (format) {
+    case PIXMAN_a8r8g8b8:
+        shm_format = WL_SHM_FORMAT_ARGB8888;
+        break;
+    case PIXMAN_x8r8g8b8:
+        shm_format = WL_SHM_FORMAT_XRGB8888;
+        break;
+    case PIXMAN_a8b8g8r8:
+        shm_format = WL_SHM_FORMAT_ABGR8888;
+        break;
+    case PIXMAN_x8b8g8r8:
+        shm_format = WL_SHM_FORMAT_XBGR8888;
+        break;
+    default:
+        ivi_screenshot_send_error(l->screenshot,
+                                  IVI_SCREENSHOT_ERROR_NOT_SUPPORTED,
+                                  "unsupported pixel format");
+        goto err_readpix;
```

Comment:
It is better to control this in the beginning of the function. If pixelformat is not supported, you can return immediately